Source/santad/SNTDriverManager.m

```diff
@@ -86,6 +70,61 @@ - (void)dealloc {
   IOServiceClose(_connection);
 }
 
+#pragma mark Driver Waiting
+
+// Helper function used with IOServiceAddMatchingNotification which expects
+// a DriverAppearedBlock to be passed as the 'reference' argument. The block
+// will be called for each object iterated over and if the block wants to keep
+// the object, it should call IOObjectRetain().
+typedef void (^DriverAppearedBlock)(io_object_t object);
+static void driverAppearedHandler(void *info, io_iterator_t iterator) {
+  DriverAppearedBlock block = (__bridge DriverAppearedBlock)info;
+  io_object_t object = 0;
+  while ((object = IOIteratorNext(iterator))) {
+    block(object);
+    IOObjectRelease(object);
+  }
+}
+
+// Wait for the driver to
+- (io_service_t)waitForDriver:(CFDictionaryRef CF_RELEASES_ARGUMENT)matchingDict {
+  IONotificationPortRef notificationPort = IONotificationPortCreate(kIOMasterPortDefault);
+  CFRunLoopAddSource([[NSRunLoop currentRunLoop] getCFRunLoop],
+                     IONotificationPortGetRunLoopSource(notificationPort),
+                     kCFRunLoopDefaultMode);
+
+  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+  io_iterator_t iterator = 0;
+  __block io_service_t service = 0;
+
+  DriverAppearedBlock block = ^(io_object_t object) {
+    IOObjectRetain(object);
+    service = (io_service_t)object;
+    dispatch_semaphore_signal(sema);
+  };
+
+  LOGI(@"Waiting for Santa driver to become available");
+
+  IOServiceAddMatchingNotification(notificationPort,
+                                   kIOMatchedNotification,
+                                   matchingDict,
+                                   driverAppearedHandler,
+                                   (__bridge_retained void *)block,
```

Comment:
Does the block fire on a different thread? Just want to make sure the below dispatch wait won’t block the execution of the block due to notification.

Reply:
It did, I've changed the scope of the PR.